collectors/node/cpu.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2016 Mesosphere, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"math"
+	"sync"
+
+	"github.com/dcos/dcos-metrics/producers"
+	"github.com/shirou/gopsutil/cpu"
+)
+
+type cpuCoresMetric struct {
+	cpuCores  int32
+	cpuTotal  float64
+	cpuUser   float64
+	cpuSystem float64
+	cpuIdle   float64
+	cpuWait   float64
+	timestamp string
+}
+
+func (m *cpuCoresMetric) poll() error {
+	ts := thisTime()
+	hi, err := getNumCores()
+	if err != nil {
+		return err
+	}
+
+	cpuStatePcts := calculatePcts(getCPUTimes())
+
+	m.cpuCores = hi
+	m.timestamp = ts
+	m.cpuTotal = cpuStatePcts.User + cpuStatePcts.System
+	m.cpuUser = cpuStatePcts.User
+	m.cpuSystem = cpuStatePcts.System
+	m.cpuIdle = cpuStatePcts.Idle
+	m.cpuWait = cpuStatePcts.Iowait
+
+	return nil
+}
+
+func (m *cpuCoresMetric) addDatapoints(nc *nodeCollector) error {
+	cpuDps := []producers.Datapoint{
+		producers.Datapoint{
+			Name:      CPU_CORES,
+			Unit:      COUNT,
+			Value:     m.cpuCores,
+			Timestamp: m.timestamp,
+		},
+		producers.Datapoint{
+			Name:      CPU_TOTAL,
+			Unit:      COUNT,
+			Value:     m.cpuTotal,
+			Timestamp: m.timestamp,
+		},
+		producers.Datapoint{
+			Name:      CPU_USER,
+			Unit:      COUNT,
+			Value:     m.cpuUser,
+			Timestamp: m.timestamp,
+		},
+		producers.Datapoint{
+			Name:      CPU_SYSTEM,
+			Unit:      COUNT,
+			Value:     m.cpuSystem,
+			Timestamp: m.timestamp,
+		},
+		producers.Datapoint{
+			Name:      CPU_IDLE,
+			Unit:      COUNT,
+			Value:     m.cpuIdle,
+			Timestamp: m.timestamp,
+		},
+		producers.Datapoint{
+			Name:      CPU_WAIT,
+			Unit:      COUNT,
+			Value:     m.cpuWait,
+			Timestamp: m.timestamp,
+		},
+	}
+
+	for _, dp := range cpuDps {
+		nc.datapoints = append(nc.datapoints, dp)
+	}
+
+	return nil
+}
+
+// -- helpers
+func getNumCores() (int32, error) {
+	cores := int32(0)
+	cpus, err := cpu.Info()
+	if err != nil {
+		return cores, err
+	}
+	for _, c := range cpus {
+		cores += c.Cores
+	}
+	return cores, nil
+}
+
+/* CPU Helper Methods */
+type lastCPUTimes struct {
+	sync.Mutex
+	times cpu.TimesStat
+}
+
+var lastCPU lastCPUTimes
+
+func init() {
```

Comment:
it's used by two funcs, I tried to refactor that out. I'm using the legacy code in this instance.